### PR TITLE
add doc for generic types & HashedSet

### PR DESCRIPTION
**Pull request by freedomhero**

### CODE REVIEW

File: docs/contracts.rst [hunk 1]
```diff
@@ -221,6 +221,71 @@ Most functions of `HashedMap` require not only a key, but also its index, ranked
...
+Most functions of `HashedSet` require an index, ranked by the value's sha256 hash in ascending order.
+
+**Constructor**
+
+* ``HashedSet(bytes data)``
+  Create an instance of ``HashedSet`` with some initial data.
+
+    .. code-block:: solidity
+
+        HashedSet<bytes, int> set = new HashedSet<bytes, int>(b'');
```
Comment (xhliu) [inline docs/contracts.rst:238]: Only takes one type, not 2. Change everywhere.
`HashedSet<int> set = new HashedSet<int>(b'');`

File: docs/contracts.rst [hunk 2]
```diff
@@ -221,6 +221,71 @@ Most functions of `HashedMap` require not only a key, but also its index, ranked
...
+        // key and value types can be omitted
+        HashedSet<int, bool> set1 = new HashedSet(b'');
+        // key and value types cannot be omitted since they cannot be inferred
+        auto set2 = new HashedSet<int, int>(b'');
+
+**Instance methods**
+
+* ``add(V val, int index) : bool``
```
Comment (xhliu) [inline docs/contracts.rst:246]: V val -> K key. Everywhere

**Review by xhliu — CHANGES_REQUESTED** [review]
Let's also add `clear()` for HashedMap.

File: docs/syntax.rst [hunk 3]
```diff
@@ -168,6 +168,35 @@ Type aliases create a new name for a type. It does not actually create a new typ
         type Age = int;
         type Coordinate = int[2];
 
+Generics/Generic Types
```
Comment (xhliu) [inline docs/syntax.rst:171]: Two more things:

1. add `flattenSha256(A a) : Sha256` into https://scryptdoc.readthedocs.io/en/latest/functions.html#hashing
2. Let us put all types into a separate file/subpage `types.rst`, now it grows big.

Comment (freedomhero) [reply docs/syntax.rst:171]: 2. The whole structure need some adjustments, let's do it in  https://github.com/sCrypt-Inc/specification/issues/10

File: docs/contracts.rst [hunk 4]
```diff
@@ -221,6 +221,76 @@ Most functions of `HashedMap` require not only a key, but also its index, ranked
...
+        require(set.has(b'1234', 0));
+
+* ``delete(E entry, int index) : bool``
+  Delete the entry with given `entry` and the index is `index`. Returns `true` if successful; otherwise returns `false`.
+
+    .. code-block:: solidity
+
+        require(set.delete(b'1234', 0));
+
+* ``clear() : bool``
```
Comment (xhliu) [inline docs/contracts.rst:272]: add clear() to HashedMap as well.

File: docs/functions.rst [hunk 5]
```diff
@@ -123,6 +123,13 @@ Hashing
 
   sha256(sha256(b))
 
+* ``Sha256 flattenSha256(T a)``
+
+  Returns a Sha256 bytes for the given argument ``a`` of any type. ``T`` here is a generic type which can be any basic type or user-defined struct type.
```
Comment (xhliu) [inline docs/functions.rst:128]: Returns Sha256 for the given argument ``a`` of any type.
  If `T` is a basic type, like ``bool`` / ``int`` / ``bytes``,  the return is the same as ``sha256(a)``.
  If `T` is a composite type (i.e., array and struct), it concatenates all the sha256 values for each flattened fields of `a` to form a joint bytes, and then call `sha256` on it to get the final result.

**Review by xhliu — APPROVED** [review]
👍 

Some minor changes before merging.

**Pull request merged**